### ISSUE TITLE
Output top-level ref definition when converting from TypeScript to JSON Schema?

### ISSUE THREAD

**Issue by sideshowbarker** [glideapps/quicktype#1519]
Is it possible to make `quicktype` output a top-level ref definition when converting from TypeScript to JSON Schema?

When I try a conversion with `quicktype`, it doesn’t output any such top-level ref definition in the generated schema — and without that included, validation tools will unexpectedly claim any JSON you check against that schema is valid, no matter whether that JSON matches the actual substantive definition in the schema.

With the `typescript-json-schema` tool, I can use the `--topRef` option to ensure it emits a top-level ref definition in the output. (See the description of that option at https://github.com/YousefED/typescript-json-schema#command-line).

And the `ts-json-schema-generator` tool seems to output a top-level ref definition by default.

---

For the sake of clarity, here’s what I mean:

Given the following TypeScript input:
```typescript
export interface Foo {
  baz: string;
}
```
…if I process that with `typescript-json-schema --topRef --required Foo.ts Foo`, I get the following output:
```json
{
    "$ref": "#/definitions/Foo",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": {
        "Foo": {
            "properties": { "baz": { "type": "string" } },
            "required": [ "baz" ],
            "type": "object"
        }
    }
}
```
Note the `"$ref": "#/definitions/Foo"` top-level ref definition.

But if I try the same TypeScript source with `quicktype --src-lang typescript --lang schema Foo.ts`, I get this:
```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": {
        "Foo": {
            "title": "Foo",
            "type": "object",
            "properties": { "baz": { "type": "string", "title": "baz" } },
            "required": [ "baz" ]
        }
    }
}
```
Note the lack of any `"$ref": "#/definitions/Foo"` top-level ref definition.

Without that top-level ref definition in the output, there’s nothing that’s referencing the necessary schema definition — and so when trying to use the schema to do validation, the schema’s not actually causing any validation to be applied.

And so when using a validation tool with that schema, no matter what JSON document I try to check with that schema, even JSON documents that are invalid according to that `Foo` schema definition, the validation tool will claim that the JSON document passes validation — because no actual validation is getting done, due to the lack of any top-level ref definition that references the `Foo` definition.

**Comment by jakeklassen** [comment]
@sideshowbarker did you ever find a solution to this within quicktype? I'm running into the same issue and find `typescript-json-schema` pretty slow in comparison.

**Comment by sideshowbarker** [comment]
> @sideshowbarker did you ever find a solution to this within quicktype?

Nope. Never got any response at all, and never came up with any workaround myself

> I'm running into the same issue and find `typescript-json-schema` pretty slow in comparison.

Yeah I can vaguely recall there were other differences that made me prefer quicktype, but I can’t recall the details now. I’ve not really messed around with any of it further since the time when I first set it up all (around when I raised this issue)
